driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -92,51 +85,67 @@ impl StableXOrderBookReading for UpdatingOrderbook {
-    stream: impl Stream<Item = Result<Event<batch_exchange::Event>, ExecutionError>>,
 ) -> Result<()> {
-    // `select!` requires the futures to be fused...
-    let exit_indicator = exit_indicator.fuse();
-    let past_events = past_events.fuse();
-    let stream = stream.fuse();
-    // ...and pinned.
-    pin_mut!(exit_indicator);
-    pin_mut!(past_events);
-    pin_mut!(stream);
+    const POLL_INTERVALL: Duration = Duration::from_secs(15);
+    const BLOCK_RANGE: u64 = 25;
 
     log::info!("Starting event based orderbook updating.");
 
+    // `select!` requires the futures to be fused and pinned.
+    let exit_indicator = exit_indicator.fuse();
+    pin_mut!(exit_indicator);
+
+    let mut last_handled_block = 0u64;
     loop {
-        // We select over everything together instead of for example the past events first then the
-        // stream to ensure that the stream gets polled at least once which it needs in order to
-        // create the corresponding filter on the node.
+        let current_block = web3.eth().block_number().compat().await?;
+        let from_block = last_handled_block.saturating_sub(BLOCK_RANGE);
+        let to_block = BlockNumber::Number(current_block);
         select_biased! {
             _ = exit_indicator => return Ok(()),
-            event = stream.next() => {
-                log::info!("Received new event.");
-                let event = event.ok_or(anyhow!("stream ended"))??;
-                handle_event(&orderbook, &mut block_timestamp_reader, event).await?;
-            },
-            past_events = past_events => {
-                let past_events = past_events?;
-                log::info!("Received {} past events.", past_events.len());
-                for event in past_events {
-                    handle_event(&orderbook, &mut block_timestamp_reader, event).await?;
-                }
-                log::info!("Finished applying past events");
-                orderbook_ready.store(true, Ordering::SeqCst);
+            events = contract.past_events(BlockNumber::Number(from_block.into()), to_block).fuse() => {
+                handle_events(&orderbook, &mut block_timestamp_reader, events?, from_block).await?;
             },
         };
+        last_handled_block = current_block.as_u64();
+        // This will be set and stay true the first time we reach this point.
+        orderbook_ready.store(true, Ordering::SeqCst);
+        // TODO: This is not optimal because it means we might sleep for some time when exit
+        // indicator triggered. We should create a sleep future that expires after poll interval and
+        // join both together.
+        std::thread::sleep(POLL_INTERVALL);
     }
 }
 
+/// Apply a vector of events to the orderbook.
+async fn handle_events(
+    orderbook: &Mutex<Orderbook>,
+    block_timestamp_reader: &mut impl BlockTimestampReading,
+    events: Vec<Event<batch_exchange::Event>>,
```

Comment:
cool, now that there is one function for this, we probably want to prepare the timestamp cache here.